### PR TITLE
Add frame extraction tool for video processing

### PR DESCRIPTION
**Pull request by humpydonkey**

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -84,4 +88,5 @@ module = [
     "faiss.*",
     "openai.*",
     "sentence_transformers.*",
+    "moviepy.*",
```
Comment (dillonalaird) [inline pyproject.toml:91]: if it's not a big requirement, we can just add it to the main dependencies. See my comment below, not sure if it'll fail anyways when it tries to import moviepy

File: vision_agent/tools/tools.py [hunk 2]
```diff
@@ -457,6 +457,42 @@ def __call__(self, input: List[int]) -> float:
...
+    def __call__(self, video_uri: str) -> list[tuple[str, float]]:
+        try:
+            from vision_agent.tools.video import extract_frames_from_video
+        except Exception as e:
+            raise ImportError(
+                "vision_agent is not installed correctly (cause: missing dependencies), please run 'pip install vision-agent[video]' instead."
+            ) from e
+        frames = extract_frames_from_video(video_uri)
+        result = []
+        _LOGGER.info(
```
Comment (dillonalaird) [inline vision_agent/tools/tools.py:486]: more of an FYI I've been using the _LOGGER.info for printing in verbose mode, but not sure if that's the best way to do it (maybe debug would have been better?)

**Review by dillonalaird — CHANGES_REQUESTED** [review]
Mostly just FYIs, I'd probably just delete the last if __name__ part and then it looks good

File: tests/tools/test_video.py [hunk 3]
```diff
@@ -0,0 +1,7 @@
+from vision_agent.tools.video import extract_frames_from_video
+
+
+def test_extract_frames_from_video():
+    video_path = "tests/data/video/test.mp4"
```
[the remark this answers is not in the document]
Comment (humpydonkey) [reply tests/tools/test_video.py:5]: good idea, i added a TODO.
generating a video isn't too bad, 
just need to make sure the CI environment has the correct libs/codec to encode a video.

File: pyproject.toml (hunk 1 above)
Comment (humpydonkey) [reply pyproject.toml:91]: good idea, love it, done
below are the required dependencies of `moviepy`, it's not huge:
```
    "decorator>=4.0.2,<6.0",
    "imageio>=2.5,<3.0",
    "imageio_ffmpeg>=0.2.0",
    "numpy>=1.17.3",
    "proglog<=1.0.0",
```

File: vision_agent/tools/tools.py (hunk 2 above)
Comment (humpydonkey) [reply vision_agent/tools/tools.py:486]: ah, it depends on how verbose we want to be.
Some third party library could be very verbose and flood our logs if we allow debug level.
but we could also selectively turn on debug mode for certain modules. 
Anyway, i think we can tune these settings later down the road.

File: vision_agent/tools/video.py [hunk 4]
```diff
@@ -0,0 +1,190 @@
+import logging
+import math
+import os
+from concurrent.futures import ProcessPoolExecutor, as_completed
+from typing import cast
+
+import cv2
+import numpy as np
+from moviepy.video.io.VideoFileClip import VideoFileClip
```
[the remark this answers is not in the document]
Comment (humpydonkey) [reply vision_agent/tools/video.py:9]: it won't get automatically imported unless someone reference it.

i like your idea below to make `moviepy` part of the main dependency group.

below are the required dependencies of `moviepy`, it's not huge:
```
    "decorator>=4.0.2,<6.0",
    "imageio>=2.5,<3.0",
    "imageio_ffmpeg>=0.2.0",
    "numpy>=1.17.3",
    "proglog<=1.0.0",
```

File: vision_agent/tools/tools.py [hunk 5]
```diff
@@ -457,6 +457,42 @@ def __call__(self, input: List[int]) -> float:
...
+                "parameters": {"video_uri": "www.foobar.com/video?name=test.mp4"},
+            },
+            {
+                "scenario": "Can you extract the images from this video file? Video path: tests/data/test.mp4",
+                "parameters": {"video_uri": "tests/data/test.mp4"},
+            },
+        ],
+    }
+
+    def __call__(self, video_uri: str) -> list[tuple[str, float]]:
```
[the remark this answers is not in the document]
Comment (humpydonkey) [reply vision_agent/tools/tools.py:526]: sounds good, btw, our va lib only supports py 3.10 and 3.11 ATM.

File: vision_agent/tools/video.py [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+    )
+    _LOGGER.debug(f"Image diff: {change_percentage}")
+    return change_percentage < threshold
+
+
+# res = extract_frames(video)
+if __name__ == "__main__":
+    video_path = "/Users/asia/Downloads/frames/baby_cam1.MP4"
+    res = extract_frames_from_video(video_path)
+    print("done, extracted num frames: ", len(res))
```
[the remark this answers is not in the document]
Comment (humpydonkey) [reply vision_agent/tools/video.py]: done

File: vision_agent/tools/video.py [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+) -> list[tuple[np.ndarray, float]]:
+    """Extract frames from a video
+
+    Parameters
+    ----------
+    video_uri: str, the path to the video file or a video file url
+    fps: int, the frame rate per second to extract the frames
+    motion_detection_threshold: float, the threshold to detect the motion between frames.
+        A value between 0-1, the percentage change that is considered a different frame.
+        A lower value means more frames will be extracted.
```
[the remark this answers is not in the document]
Comment (humpydonkey) [reply vision_agent/tools/video.py]: yeah, it is.
sg, updated it.

**Review by dillonalaird — APPROVED** [review]
LGTM

**Pull request merged**